File: app/src/main/java/com/musicplayer/SocyMusic/data/base/PlaylistSong.java
package com.musicplayer.SocyMusic.data.base;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Ignore;

import com.musicplayer.SocyMusic.data.Playlist;
import com.musicplayer.SocyMusic.data.Song;

import java.util.UUID;

@Entity(tableName = "Playlist_Song", primaryKeys = {"playlist_id", "song_id", "song_index"}
        , foreignKeys = {@ForeignKey(entity = Playlist.class,
        parentColumns = "playlist_id",
        childColumns = "playlist_id",
        onDelete = ForeignKey.CASCADE)
        , @ForeignKey(entity = Song.class,
        parentColumns = "song_id",
        childColumns = "song_id",
        onDelete = ForeignKey.CASCADE)})
public class PlaylistSong {
    @NonNull
    @ColumnInfo(name = "playlist_id")
    public final UUID playlistID;
    @NonNull
    @ColumnInfo(name = "song_id")
    public final UUID songID;
    @ColumnInfo(name = "song_index")
    public int index;

    @Ignore
    public PlaylistSong(@NonNull UUID playlistID, @NonNull UUID songID) {
        this.playlistID = playlistID;
        this.songID = songID;
    }

    public PlaylistSong(@NonNull UUID playlistID, @NonNull UUID songID, int index) {
        this(playlistID, songID);
        this.index = index;
    }
}

File: app/src/main/java/com/musicplayer/OpenMusic/data/base/PlaylistSong.kt
package com.musicplayer.OpenMusic.data.base

import com.musicplayer.OpenMusic.data.Playlist
import com.musicplayer.OpenMusic.data.Song
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.ForeignKey.Companion.CASCADE
import androidx.room.Ignore
import java.util.*

@Entity(
    tableName = "Playlist_Song",
    primaryKeys = ["playlist_id", "song_id", "song_index"],
    foreignKeys = [ForeignKey(
        entity = Playlist::class,
        parentColumns = arrayOf("playlist_id"),
        childColumns = arrayOf("playlist_id"),
        onDelete = CASCADE
    ), ForeignKey(
        entity = Song::class,
        parentColumns = arrayOf("song_id"),
        childColumns = arrayOf("song_id"),
        onDelete = CASCADE
    )]
)
class PlaylistSong @Ignore constructor(
    @field:ColumnInfo(name = "playlist_id") val playlistID: UUID, @field:ColumnInfo(
        name = "song_id",
        index = true
    ) val songID: UUID
) {
    @JvmField
    @ColumnInfo(name = "song_index")
    var index = 0

    constructor(playlistID: UUID, songID: UUID, index: Int) : this(playlistID, songID) {
        this.index = index
    }
}